Error from InMemDB: Error: ERROR setObject id xxx The database is not open!

Habe gerade einmal ein update auf die 4.0.3 gemacht.
Gestern abend hatte ich noch ein apt update / upgrade gemacht, wobei auch nodejs aktualisiert wurde. Aktuelle node und npm Version:
```
proxmox@ioBrokerTestsystem:/opt/iobroker$ node -v
v14.19.0
proxmox@ioBrokerTestsystem:/opt/iobroker$ npm -v
6.14.16
```

Beim Update gab es diese Fehlermeldungen. Haben sie was zu bedeuten?
Die IP 192.168.178.78 ist der ioBroker Slave. Ich habe gerade den Master aktualisiert, der Slave (js-controller 4.0 alpha 73) lief aber.

```
Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.cputime: The database is not open!
Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.mem: The database is not open!
Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.memHeapUsed: The database is not open!
```

Hier die vollständie Ausgabe vom Update
iobrokernew ist der Slave mit der IP 192.168.178.78. Das update lief auf dem Master "ioBrokerTestsystem" mit der IP 192.168.178.77

```
proxmox@ioBrokerTestsystem:/opt/iobroker$ npm i iobroker.js-controller@next

	
[sudo] password for proxmox:

	
 

	
> iobroker.js-controller@4.0.3 preinstall /opt/iobroker/node_modules/iobroker.js-controller

	
> node lib/preinstallCheck.js

	
 

	
NPM version: 6.14.16

	
 

	
> unix-dgram@2.0.4 install /opt/iobroker/node_modules/unix-dgram

	
> node-gyp rebuild

	
 

	
make: Entering directory '/opt/iobroker/node_modules/unix-dgram/build'

	
  CXX(target) Release/obj.target/unix_dgram/src/unix_dgram.o

	
In file included from ../src/unix_dgram.cc:5:

	
../../nan/nan.h: In function 'void Nan::AsyncQueueWorker(Nan::AsyncWorker*)':

	
../../nan/nan.h:2294:62: warning: cast between incompatible function types from 'void (*)(uv_work_t*)' {aka 'void (*)(uv_work_s*)'} to 'uv_after_work_cb' {aka 'void (*)(uv_work_s*, int)'} [-Wcast-function-type]

	
 2294 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)

	
      |                                                              ^

	
In file included from ../../nan/nan.h:56,

	
                 from ../src/unix_dgram.cc:5:

	
../src/unix_dgram.cc: At global scope:

	
/home/iobroker/.cache/node-gyp/14.19.0/include/node/node.h:793:43: warning: cast between incompatible function types from 'void (*)(v8::Local<v8::Object>)' to 'node::addon_register_func' {aka 'void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)'} [-Wcast-function-type]

	
  793 |       (node::addon_register_func) (regfunc),                          \

	
      |                                           ^

	
/home/iobroker/.cache/node-gyp/14.19.0/include/node/node.h:827:3: note: in expansion of macro 'NODE_MODULE_X'

	
  827 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)

	
      |   ^~~~~~~~~~~~~

	
../src/unix_dgram.cc:404:1: note: in expansion of macro 'NODE_MODULE'

	
  404 | NODE_MODULE(unix_dgram, Initialize)

	
      | ^~~~~~~~~~~

	
  SOLINK_MODULE(target) Release/obj.target/unix_dgram.node

	
  COPY Release/unix_dgram.node

	
make: Leaving directory '/opt/iobroker/node_modules/unix-dgram/build'

	
 

	
> iobroker.js-controller@4.0.3 install /opt/iobroker/node_modules/iobroker.js-controller

	
> node iobroker.js setup first

	
 

	
object _design/system updated

	
{

	
  "defaultPrivate": "-----BEGIN RSA PRIVATE KEY-----\r\nMIIEpQIBAAKCAQEA0UJUHJOm8DL4S++hcAHHj4DLTZWBTzRQufvVYL0FdT5qGti0\r\nbCmu5KdFCL0uwOT89EEGE2rp9kwyneeeqaBXpSoScX78gtX7wOUTUX+pyBQi82kR\r\nsFX/WYrO8o4h1bEuOqyTGD4sywE1gupGnuAytgimuARGKB4i9wLbIM3k22XjCmnW\r\nRgp9VSNOUE8HhCKKHHi4TUs0dBaH2zdacKoMsiqB5OZw3v+VfV+HlG7edraTvxnI\r\nOd7FIjhb1ed5DFbFLKHrKpA2snSkZuMJvReE/5vvy1awnQOpOsJB0N6fpO/nRCt6\r\n0HolCiH5lwAkTVlxieQowsW9cGIUHCwCi0BhXQIDAQABAoIBAQDHxP+RbG/xXWdm\r\n7vR6fmhqGfV8Mp8I4xOyCbWxraIZRKV8zaaon76EMR+ZcSnCIplOAQlw3h9O731e\r\n+7CflSA24AxNZ4Je0TOO+eHQawTFqluXbpXhdqfYbsiUDJZGb+iHQRsICMswXg4c\r\n1dTndXdUCrxGbmYfhG2yzHovPgAeZdNHG15h9HAnpxT6PcZiPAapnQCk3Yx7n9zs\r\nhgcXASBGL2DrOXS9qiQ6Tb9AWVDiM39mJX8yE0utNNXa8dgtppOhyu6xcBDyvMbM\r\nf2uBFMZlWU2FGa901A8gKRJWN+WItC7HxCJYtWD2fcXCQcMS7aYATY8rm0Rre3q6\r\no9niOoXNAoGBAOeogYRhMsZ+DkF5Rh7JgznIbTFEuEMYAXJaWNz4ZoyqGC7qRoft\r\nRlv+nj5YGb2La90qG+jblPjUvGW8ngdodmj9+pI9UFzt/Voo0Zskhh9BXS5Gb7JA\r\n+ieUPy0mLwXnBDwhW3FwSEVkLIkabGlrrRdg+5mcasSkSbNaX34bkPG7AoGBAOc/\r\nTABAANs0ACzzXwNh8LUKNft8Lm18dG+OmNu/Xn8OcZ/hgP84HUKPhN13h5IXLrPo\r\nTudvVRzc7B2kbYpwLJNrVge3ZRV9yUtoUICNdjnurR+DV6+nGA+byF08mtLdQJH5\r\n3vmhoZE8D7cvBR+IEsf3T2MKRqkLy8FNHtQtBVvHAoGBAKv2iRZEtNVls51kseme\r\nutvE7VF6arHbew9xA7AccLuQdF03KV4zg1w82Sj8azcp7MsDPC6R+bbw7F01M7Xa\r\nFLQBfOTbwPEK9VvdGyNobGt9MPdt7/JJi6ifwnn4c0WVZK5iMKg3ql5gazLnO3zS\r\nIm5NXtd7pkIMlm0d8+7/9YuLAoGAVe/E67bdWpjpk8rhvV0RWgYwZ4QPsOp2Ttrc\r\nXZWxpWAVw8fG65HL9edmBfOcsyLgXUVwxTsvkylaqe1oM/FZ73MNJpL2AtWbD+X1\r\nPMDVddQIDFUAKivbmjlXbabPFpvYmicUwQpbA6Ehaf8UVVJ5QEzc+L5h4gKbNvvV\r\nAB63BhMCgYEAwrHBgWtdCzrEg9c663uziK91j6T4PKHrua1yka4awT9Z8LhEga24\r\nMhXED9Bw9Vvsa6xDaJ7AobGEtMTYCupB2u+cemaR3/f7ayCNBlZuTjz/ibftHkJb\r\nV9blMZ0EBC/006ZLnsA6QLs0SqruVgPQbUjLP5Q7F4Dmudz9swxpuT0=\r\n-----END RSA PRIVATE KEY-----\r\n",

	
  "defaultPublic": "-----BEGIN CERTIFICATE-----\r\nMIIDlDCCAnygAwIBAgIJCEBGN+fDGlYlMA0GCSqGSIb3DQEBCwUAMD4xETAPBgNV\r\nBAMTCGlvYnJva2VyMRYwFAYDVQQKEw1pb0Jyb2tlciBHbWJIMREwDwYDVQQLEwhp\r\nb2Jyb2tlcjAeFw0yMjAyMDUxMDM5MTVaFw0yMzAyMDUxMDM5MTVaMEgxGzAZBgNV\r\nBAMTEmlvQnJva2VyVGVzdHN5c3RlbTEWMBQGA1UEChMNaW9Ccm9rZXIgR21iSDER\r\nMA8GA1UECxMIaW9icm9rZXIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIB\r\nAQDRQlQck6bwMvhL76FwAcePgMtNlYFPNFC5+9VgvQV1Pmoa2LRsKa7kp0UIvS7A\r\n5Pz0QQYTaun2TDKd556poFelKhJxfvyC1fvA5RNRf6nIFCLzaRGwVf9Zis7yjiHV\r\nsS46rJMYPizLATWC6kae4DK2CKa4BEYoHiL3AtsgzeTbZeMKadZGCn1VI05QTweE\r\nIooceLhNSzR0FofbN1pwqgyyKoHk5nDe/5V9X4eUbt52tpO/Gcg53sUiOFvV53kM\r\nVsUsoesqkDaydKRm4wm9F4T/m+/LVrCdA6k6wkHQ3p+k7+dEK3rQeiUKIfmXACRN\r\nWXGJ5CjCxb1wYhQcLAKLQGFdAgMBAAGjgYowgYcwDAYDVR0TAQH/BAIwADAPBgNV\r\nHQ8BAf8EBQMDB7+AMB0GA1UdEQQWMBSCEmlvQnJva2VyVGVzdHN5c3RlbTAdBgNV\r\nHQ4EFgQUD0SE7Kg2dRW/OQU/gUV8P97Rbj4wHQYDVR0lBBYwFAYIKwYBBQUHAwEG\r\nCCsGAQUFBwMCMAkGA1UdIwQCMAAwDQYJKoZIhvcNAQELBQADggEBANBTXcSPdI7E\r\niwmsf7h5eubmtIeGNU0UYIYiVvqgBn0DUw8SwjwhZgbFCPd4Qx6NAecEPWqztmem\r\n2HVqwkzgpzQ64wb1YOywZqUP7xmWlnxzEzUZg3hq8+JGgJqkl3tMTnpm87jGReyu\r\nYjfyRHtNj4Gjq6LGs8RqPA4pXe204XSoakdKoeRiy7qi2ULD8GKLjkC+pHF4RA/Y\r\nKPGmY0F8RliFYPkdIvN8GbCpibZnwq+J6U+1+9k46EuXhzaLQVjGfCNLPzjmmT/0\r\nYtosTkCOIbIwiwmmP9byXHhimVXOq/97cvR+zZB+xxqJXvt5zxTGjKgNAcpN/w7E\r\n0F41Cw5EmdU=\r\n-----END CERTIFICATE-----\r\n"

}

Update certificate defaultPrivate

The object "system.certificates" was updated successfully.

Update certificate defaultPublic

The object "system.certificates" was updated successfully.

Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.cputime: The database is not open!

Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.mem: The database is not open!

Server Objects 192.168.178.78:60268 Error from InMemDB: Error: ERROR setObject id=system.host.ioBrokerNew.memHeapUsed: The database is not open!

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/chokidar/node_modules/fsevents):

npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: osx-temperature-sensor@1.0.7 (node_modules/osx-temperature-sensor):

npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for osx-temperature-sensor@1.0.7: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

 

+ iobroker.js-controller@4.0.3

removed 19 packages and updated 14 packages in 102.06s

 

43 packages are looking for funding

  run `npm fund` for details
````

--- a/packages/common/package.json
+++ b/packages/common/package.json
@@ -6,7 +6,7 @@
   },
   "optionalDependencies": {
     "@datalust/winston-seq": "^1.0.0",
-    "winston-syslog": "^2.4.4"
+    "winston-syslog": "^2.5.0"
   },
   "dependencies": {
     "@alcalzone/pak": "^0.7.0",

--- a/packages/controller/main.js
+++ b/packages/controller/main.js
@@ -603,7 +603,7 @@ function createObjects(onConnect) {
                 logger.error(`${hostLogPrefix} Cannot subscribe to primary host expiration: ${e.message}`);
             }
 
-            if (!primaryHostInterval) {
+            if (!primaryHostInterval && !compactGroupController) {
                 primaryHostInterval = setInterval(checkPrimaryHost, PRIMARY_HOST_LOCK_TIME / 2);
             }
 
@@ -901,7 +901,7 @@ function startAliveInterval() {
  */
 async function checkPrimaryHost() {
     // we cannot interact with db now because currently reconnecting
-    if (objectsDisconnectTimeout) {
+    if (objectsDisconnectTimeout || compactGroupController) {
         return;
     }
 

--- a/packages/db-objects-file/lib/objects/objectsInMemServerRedis.js
+++ b/packages/db-objects-file/lib/objects/objectsInMemServerRedis.js
@@ -45,8 +45,8 @@ const ObjectsInMemoryFileDB = require('./objectsInMemFileDB');
 //
 
 /**
- * This class inherits statesInMemoryFileDB class and adds socket.io communication layer
- * to access the methods via socket.io
+ * This class inherits statesInMemoryFileDB class and adds redis communication layer
+ * to access the methods via redis protocol
  **/
 class ObjectsInMemoryServer extends ObjectsInMemoryFileDB {
     /**
@@ -843,28 +843,26 @@ class ObjectsInMemoryServer extends ObjectsInMemoryFileDB {
      * Destructor of the class. Called by shutting down.
      */
     async destroy() {
-        await super.destroy();
-
         if (this.server) {
             Object.keys(this.serverConnections).forEach(s => {
                 this.serverConnections[s].close();
                 delete this.serverConnections[s];
             });
 
-            return /** @type {Promise<void>} */ (
-                new Promise(resolve => {
-                    if (!this.server) {
-                        return void resolve();
-                    }
-                    try {
-                        this.server.close(() => resolve());
-                    } catch (e) {
-                        console.log(e.message);
-                        resolve();
-                    }
-                })
-            );
+            await new Promise(resolve => {
+                if (!this.server) {
+                    return void resolve();
+                }
+                try {
+                    this.server.close(() => resolve());
+                } catch (e) {
+                    console.log(e.message);
+                    resolve();
+                }
+            });
         }
+
+        await super.destroy();
     }
 
     /**

--- a/packages/db-objects-jsonl/lib/objects/objectsInMemServerRedis.js
+++ b/packages/db-objects-jsonl/lib/objects/objectsInMemServerRedis.js
@@ -45,8 +45,8 @@ const ObjectsInMemoryJsonlDB = require('./objectsInMemJsonlDB');
 //
 
 /**
- * This class inherits statesInMemoryFileDB class and adds socket.io communication layer
- * to access the methods via socket.io
+ * This class inherits statesInMemoryJsonlDB class and adds redis communication layer
+ * to access the methods via redis protocol
  **/
 class ObjectsInMemoryServer extends ObjectsInMemoryJsonlDB {
     /**
@@ -841,28 +841,25 @@ class ObjectsInMemoryServer extends ObjectsInMemoryJsonlDB {
      * Destructor of the class. Called by shutting down.
      */
     async destroy() {
-        await super.destroy();
-
         if (this.server) {
             Object.keys(this.serverConnections).forEach(s => {
                 this.serverConnections[s].close();
                 delete this.serverConnections[s];
             });
 
-            return /** @type {Promise<void>} */ (
-                new Promise(resolve => {
-                    if (!this.server) {
-                        return void resolve();
-                    }
-                    try {
-                        this.server.close(() => resolve());
-                    } catch (e) {
-                        console.log(e.message);
-                        resolve();
-                    }
-                })
-            );
+            await new Promise(resolve => {
+                if (!this.server) {
+                    return void resolve();
+                }
+                try {
+                    this.server.close(() => resolve());
+                } catch (e) {
+                    console.log(e.message);
+                    resolve();
+                }
+            });
         }
+        await super.destroy();
     }
 
     /**
@@ -977,7 +974,7 @@ class ObjectsInMemoryServer extends ObjectsInMemoryJsonlDB {
     _initRedisServer(settings) {
         return new Promise((resolve, reject) => {
             if (settings.secure) {
-                reject(new Error('Secure Redis unsupported for File-DB'));
+                reject(new Error('Secure Redis unsupported for JSONL-DB'));
             }
             try {
                 this.server = net.createServer();

--- a/packages/db-states-file/lib/states/statesInMemServerRedis.js
+++ b/packages/db-states-file/lib/states/statesInMemServerRedis.js
@@ -476,28 +476,25 @@ class StatesInMemoryServer extends StatesInMemoryFileDB {
      * Destructor of the class. Called by shutting down.
      */
     async destroy() {
-        await super.destroy();
-
         if (this.server) {
             for (const s of Object.keys(this.serverConnections)) {
                 this.serverConnections[s].close();
                 delete this.serverConnections[s];
             }
 
-            return /** @type {Promise<void>} */ (
-                new Promise(resolve => {
-                    if (!this.server) {
-                        return void resolve();
-                    }
-                    try {
-                        this.server.close(() => resolve());
-                    } catch (e) {
-                        console.log(e.message);
-                        resolve();
-                    }
-                })
-            );
+            await new Promise(resolve => {
+                if (!this.server) {
+                    return void resolve();
+                }
+                try {
+                    this.server.close(() => resolve());
+                } catch (e) {
+                    console.log(e.message);
+                    resolve();
+                }
+            });
         }
+        await super.destroy();
     }
 
     /**

--- a/packages/db-states-jsonl/lib/states/statesInMemServerRedis.js
+++ b/packages/db-states-jsonl/lib/states/statesInMemServerRedis.js
@@ -476,28 +476,25 @@ class StatesInMemoryServer extends StatesInMemoryJsonlDB {
      * Destructor of the class. Called by shutting down.
      */
     async destroy() {
-        await super.destroy();
-
         if (this.server) {
             for (const s of Object.keys(this.serverConnections)) {
                 this.serverConnections[s].close();
                 delete this.serverConnections[s];
             }
 
-            return /** @type {Promise<void>} */ (
-                new Promise(resolve => {
-                    if (!this.server) {
-                        return void resolve();
-                    }
-                    try {
-                        this.server.close(() => resolve());
-                    } catch (e) {
-                        console.log(e.message);
-                        resolve();
-                    }
-                })
-            );
+            await new Promise(resolve => {
+                if (!this.server) {
+                    return void resolve();
+                }
+                try {
+                    this.server.close(() => resolve());
+                } catch (e) {
+                    console.log(e.message);
+                    resolve();
+                }
+            });
         }
+        await super.destroy();
     }
 
     /**

